### DISCUSSION TITLE
[Help] Shell Commands with JS scripts

### DISCUSSION THREAD

**Discussion by FelipeRearden** [Taitava/obsidian-shellcommands#73] [Q&A] [answered]
Hello @Taitava !!!!!!!

I have no idea if SC could help me with this workflow but I decide to ask you. 
- Forgive if what I'm asinkg does not make any sense :)

## Here is my Situation 

Frequently, I import to Obsidian from iOS (more about this [here in the forum](https://forum.obsidian.md/t/ios-shortcuts-share-your-ideas/15149/78?u=felipe)) a bunch of URL that will be used inside Obsidian.
- Save some URL from artciles that I want to read/highlight/work inside Obsidian;
- URL from my Podcast App (Airr App) that generates a bunch of URL with the highlights from one episode that I'm listening. more about this workflow [here in this Git](https://github.com/trashhalo/obsidian-extract-url/issues/16)

To summarize: I have URL inside an Obsidian Note that I want to transform to .md text inside Obsidian 

## Actual Workflow to do this

To do this today I'm using the **Obsidian Web Clipper Bookmarklet** created by from our community friends - link [here](https://forum.obsidian.md/t/obsidian-web-clipper-bookmarklet-with-full-markdown-support-for-images-headings-and-code-blocks/22068)  

My attempt to explain what the boommarklet do:
- it is a `js` script that is saved as a Safari bookmark;
- it grabs all content from the URL, export to markdown and creates a new note in a specific folder (defined in the js script) in Obsidian with the content in markdown .

This is the gist for the script: https://gist.github.com/kepano/90c05f162c37cf730abb8ff027987ca3

Running this URL https://www.airr.io/quote/60153afa030ce32f95173a18 creates this new note in Obsidian

![Screen Shot 2021-10-11 at 09 11 06](https://user-images.githubusercontent.com/85180696/136789843-4ed80b23-5ae7-42bc-adeb-3524c043c6cf.png)

**Another option is the plugin EXTRACT URL CONTENT that do automatic what the Bookmarkelt does usind Safari:**
- Unfortunately it does not work well with 70% of my URL's :(

## My Help Request 

@Taitava,  do you think is it possible to use SC to use/open the `js` script with out having to go to Safari? 
- save the code as a .js file in my computer and create a SC command to execute the `.js` with the URL in the {{selection}} or {{clipboard}}

Am I crazy about asking this??????

I'm not 100% about this, but I wanna know if is something feasible :)

Have a great day and thanks for reading this :)

**Comment by Taitava** [comment, marked as answer]
Really interesting workflow! 🙂 



> @Taitava, do you think is it possible to use SC to use/open the `js` script with out having to go to Safari?
> 
>     * save the code as a .js file in my computer and create a SC command to execute the `.js` with the URL in the {{selection}} or {{clipboard}}

There's a couple of things I'm not sure about:
- Can this be done **without** opening a web browser.
- Can JavaScript files be executed directly. Sure, I guess there might be some programs that can execute JS files without opening a browser, but then again, I'm not sure if they provide the same features a web browser would provide for JavaScript, e.g. the ability to load webpages.

I think we might have two options (at least) here. I will first cover an option that does not use `obsidian-web-clipper.js` at all, but after that I will try to modify that option so that `obsidian-web-clipper.js` gets included.

# 1. option

At Least on Linux there is a command named `wget`, and I guess it can be installed on Mac, too. It downloads webpages (or other files) and outputs them to stdout - or to a file. Now, let's see how we can use it:
1. Create a new shell command: `wget {{selection}} -O -`. `-O -` tells the program to output to stdout instead of into a file.
2. Define the output channel to be *Current file: caret position*. (Or use `-O MyNote.md` or `-O MyPage.html` if you want to create a new file).
3. Now you have a really, really basic level web page downloader that replaces the currently selected URL with the webpage's content. If you don't want to lose the URL, then use `{{clipboard}}` instead of `{{selection}}` like you mentioned above.
4. BUT: This method has a huge disadvantage: it brings **everything** from the webpage to your note: navigation bars and other structures are included. The result will be ugly and requires you to manually delete stuff.
5. You can try to come up with a regular expression that would take only a subset of the downloaded HTML into your note: `wget {{selection}} | grep <body>.*?</body>` . But then again, [HTML cannot be parsed with regular expressions](https://stackoverflow.com/a/1732454/2754026) - at least not very well, so do not use too much energy for this or you will lose your mind.

# 2. option

I don't really know if this would work, and I haven't tried it myself (well, actually I did not even try the first option either). This is just pseudo code that might work, or maybe not.

1. Edit our `wget` command to store the downloaded webpage in a file: `wget {{selection}} -O temporary-file.html`.
2. We need to insert the following text into the file we just downloaded:
    ```html
    <script type="text/javascript">
        /* Put here the JavaScript content from https://gist.github.com/kepano/90c05f162c37cf730abb8ff027987ca3/raw/0406f34aa96c54bd66c02bf61ac556474345864c/obsidian-web-clipper.js 
       BUT remove the javascript: prefix. */
    </script>
    ```
4. The above HTML text (well, actually it's mostly JavaScript, but wrapped in HTML) should be inserted into `temporary-file.html`, right before a closing `</body>` tag near the end of the file. I do not know which command can do this, maybe `sed`? So I'll let it for you to find out which command does this.
5. Then we need to make our command to open `temporary-file.html` in a web browser. I don't know the exact command for Mac, but maybe `safari temporary-file.html`?
6. Combine all these commands together: `wget {{selection}} -O temporary-file.html && THE INSERT COMMAND && safari temporary-file.html`.

To be honest, I have no clue if it's possible to make this work. As I'm not familiar with `obsidian-web-clipper.js`, I'd suggest you to ask help from @kepano to edit that JavaScript code to work in this scenario. It might be that it won't work at all with my instructions.

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!!

Thank you very much for explain all of this!!!!

My hope was something that was already "integrated" with shell commands on MacOs/Win/Linux in a way that I just could run a command already made for this type o situation :(

I'm gonna follow you solution and try to talk with the creator to see if the creator can help me. 

Thanks again and I'm sorry if I open our new discussion section with a "almost impossible" user case :)

Have a great day!

**Reply by Taitava** [reply]
You are welcome! Don't be sorry! This is a really good real world example about a workflow that would be nice to achieve. It's a good opening for this section! 🙂 And who knows, maybe my answer is not the "final truth"? Maybe someone else can manage to come up with something more complete than my answer? Who knows! 😎

Have a nice day to you too! 🙂

**Reply by Taitava** [reply]
And thanks for marking my message as an answer! 🌷

**Reply by FelipeRearden** [reply]
> You are welcome! Don't be sorry! This is a really good real world example about a workflow that would be nice to achieve. 

Thank you so much @Taitava !!!! 

> And thanks for marking my message as an answer! 🌷

Sure! It's the least that I can for taking your time to help me (as you always do). 

I wish you a fantastic Wednesday!

**Reply by Taitava** [reply]
Have a nice day, too! 🙂
